src/main/java/reactor/netty/channel/MonoSendMany.java

```diff
@@ -618,6 +654,98 @@ public Void get(long timeout, TimeUnit unit) {
 			throw new UnsupportedOperationException();
 		}
 
+		// this as discard hook
+		@Override
+		public void accept(I i) {
+			try {
+				parent.sourceCleanup.accept(i);
+			}
+			catch (IllegalReferenceCountException e) {
+				// FIXME: should be removed once fusion is fixed in reactor-core
+				//        for now we have double releasing issue
+			}
+			// propagates discard to the downstream
+			Operators.onDiscard(i, actual.currentContext());
+		}
+
+		// Context interface impl
```

Comment:
this seems to be all done to avoid overhead on `get`, but now we potentially create a `Context` object _per write operation_ and _per iteration_. Plus the implementation is convoluted. Wouldn't it be saner to instantiate a `Context thisContext = actual.currentContext().put(KEY_DISCARD, this)` once at construction ?

Reply:
I don't think so. Having `actual.currentContext().put(KEY_DISCARD, this)` per operation becomes common and sometimes redundant (since the normal usage without discarding is a general case whilst discarding and releasing is a small percentage of cases. So now if we do that all the time we create a new object which decreases performance significantly (usually it is around 10% of normal throughput).

On the other hand, we have this implementation of the context which is not harmful but improves performance

Reply:
I'm not sure I followed all your argument. If discarding is a small percentage of cases, then it must mean that there are a lot of other cases where data is put in the `Context` other that discard hook. So there's a high chance something puts data in the `Context` upstream of this operator, and thus calls `MonoSendMany#put` and thus creates up to TWO intermediary `Context` objects ?

This implementation "does something redundant all the time" (in write operations).
Yes, granted, read operations are guaranteed to happen more frequently that writes.

Another issue is that every single operation on the `MonoSendMany#currentContext` will now invoke `actual.currentContext()`.

If your concern is about code that never calls `currentContext`, then have a `volatile Context` field that gets lazily initialized upon first call to `currentContext()`, wdyt ?

Reply:
I'm saying that `actual.currentContext().put(KEY_DISCARD, this)` allocates an object. If that object is stored as a field - the long-lived object will be produced, so GC will not be able to easily eliminate it.

Also, usually, context modified at the very bottom, not at the top. So it is likely that `actual.currentContext()` has values rather than later will be `MonoSendMany#put` added something.

> and thus calls MonoSendMany#put and thus creates up to TWO intermediary Context objects?

In contrast to long-lived objects, objects allocated within a stack is nothing for GC since JIT is smart enough to escape redundant objects, especially when they are immutable, so this will have at the end only a single `Context` which is identical to how it was before. 

So, the main difference of having `context` field and implementing `Context` interface is in the fact that one allocation will be avoided and no extra longlived objects will be produced.

Reply:
ok yeah, the argument that this is upstream is a good one 👍